### PR TITLE
Add nodes using terraform in vSphere

### PR DESCRIPTION
**Pull request by vavuthu**
Signed-off-by: vavuthu <vavuthu@redhat.com>

### CODE REVIEW

**Comment by vavuthu** [comment]
Verified locally

```
$ run-ci tests/manage/z_cluster/cluster_expansion/test_node_expansion.py --ocsci-conf conf/ocsci/vsphere_upi.yaml --cluster-name vavuthut-addnode4 --cluster-path ~/VJ/installations/clusterdirs/vavuthut-addnode4 --ocsci-conf ~/VJ/VMWARE/UPLOAD/new/new/vSphere-DC2_VC1.yaml
2021-02-10 20:03:56,199 - MainThread - ocs_ci.framework.pytest_customization.ocscilib - INFO - Dump of the consolidated config file is located here: /tmp/run-1612967626-config.yaml
Collecting Cluster versions
============================================================================ test session starts ============================================================================
platform linux -- Python 3.7.5, pytest-5.3.5, py-1.9.0, pluggy-0.13.1
rootdir: /home/vavuthu/VJ/Gluster/automation/commits/MASTER1/ocs-ci, inifile: pytest.ini
plugins: ordering-0.6, metadata-1.10.0, logger-0.5.1, marker-bugzilla-0.9.4, html-2.1.1, reportportal-1.10.3
collected 1 item                                                                                                                                                            

tests/manage/z_cluster/cluster_expansion/test_node_expansion.py::TestAddNode::test_add_node 
------------------------------------------------------------------------------ live log setup -------------------------------------------------------------------------------
20:04:31 - MainThread - ocs_ci.utility.utils - INFO - Retrieving the authentication config dictionary
20:04:31 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage get pods -o name
20:04:32 - MainThread - ocs_ci.ocs.utils - INFO - pod name match found appending rook-ceph-tools-67dfbfb8f5-mmqx8


20:05:18 - MainThread - ocs_ci.ocs.platform_nodes - INFO - Adding Nodes of type RHCOS
20:05:18 - MainThread - ocs_ci.ocs.platform_nodes - INFO - Existing worker nodes: 3, New nodes to add: 3
20:05:18 - MainThread - ocs_ci.utility.csr - INFO - Retrieving CSR data
20:05:18 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n default get csr  -n default -o yaml
20:05:20 - MainThread - ocs_ci.utility.utils - INFO - Repository already cloned at /home/vavuthu/VJ/Gluster/automation/commits/MASTER1/ocs-ci/external/installer, skipping clone
20:05:20 - MainThread - ocs_ci.utility.utils - INFO - Fetching latest changes from repository
20:05:20 - MainThread - ocs_ci.utility.utils - INFO - Executing command: git fetch --all
20:05:22 - MainThread - ocs_ci.utility.utils - WARNING - Command stderr: From https://github.com/openshift/installer
   2952715dc..5f28e6062  master      -> origin/master
   2952715dc..5f28e6062  release-4.8 -> origin/release-4.8

20:05:22 - MainThread - ocs_ci.utility.utils - INFO - Checking out repository to specific branch: release-4.7
20:05:22 - MainThread - ocs_ci.utility.utils - INFO - Executing command: git checkout release-4.7
20:05:22 - MainThread - ocs_ci.utility.utils - WARNING - Command stderr: Already on 'release-4.7'

20:05:22 - MainThread - ocs_ci.utility.utils - INFO - Reset branch: release-4.7 with latest changes
20:05:22 - MainThread - ocs_ci.utility.utils - INFO - Executing command: git reset --hard origin/release-4.7
20:05:22 - MainThread - root - INFO - original terraform file: /home/vavuthu/VJ/installations/clusterdirs/vavuthut-addnode4/terraform_data/terraform.tfvars_1612967722
20:05:22 - MainThread - ocs_ci.deployment.terraform - INFO - Initializing terraform work directory
20:05:22 - MainThread - ocs_ci.utility.utils - INFO - Executing command: /home/vavuthu/VJ/Gluster/automation/commits/MASTER1/ocs-ci/bin/terraform init /home/vavuthu/VJ/Gluster/automation/commits/MASTER1/ocs-ci/external/installer/upi/vsphere/
20:05:24 - MainThread - ocs_ci.utility.utils - INFO - Executing command: /home/vavuthu/VJ/Gluster/automation/commits/MASTER1/ocs-ci/bin/terraform apply -target=module.ipam_compute '-var-file=/home/vavuthu/VJ/installations/clusterdirs/vavuthut-addnode4/terraform_data/terraform.tfvars' -auto-approve  '/home/vavuthu/VJ/Gluster/automation/commits/MASTER1/ocs-ci/external/installer/upi/vsphere/'
20:05:29 - MainThread - ocs_ci.utility.utils - INFO - Executing command: /home/vavuthu/VJ/Gluster/automation/commits/MASTER1/ocs-ci/bin/terraform apply -target=module.compute_vm '-var-file=/home/vavuthu/VJ/installations/clusterdirs/vavuthut-addnode4/terraform_data/terraform.tfvars' -auto-approve  '/home/vavuthu/VJ/Gluster/automation/commits/MASTER1/ocs-ci/external/installer/upi/vsphere/'
20:09:16 - MainThread - ocs_ci.ocs.platform_nodes - INFO - Newly added compute IP's: ['10.70.60.59', '10.70.60.49', '10.70.60.60']
20:09:16 - Thread-1 - paramiko.transport - INFO - Connected (version 2.0, client OpenSSH_8.0)
20:09:17 - Thread-1 - paramiko.transport - INFO - Authentication (publickey) successful!
20:09:17 - MainThread - ocs_ci.utility.connection - INFO - Executing cmd: sudo sed -i '0,/.*:80 check$/s/.*:80 check$/        server 10.70.60.59 10.70.60.59:80 check\n&/' /etc/haproxy/haproxy.conf on 10.70.60.45
20:09:18 - MainThread - ocs_ci.utility.connection - INFO - Executing cmd: sudo sed -i '0,/.*:80 check$/s/.*:80 check$/        server 10.70.60.49 10.70.60.49:80 check\n&/' /etc/haproxy/haproxy.conf on 10.70.60.45
20:09:19 - MainThread - ocs_ci.utility.connection - INFO - Executing cmd: sudo sed -i '0,/.*:80 check$/s/.*:80 check$/        server 10.70.60.60 10.70.60.60:80 check\n&/' /etc/haproxy/haproxy.conf on 10.70.60.45
20:09:19 - MainThread - ocs_ci.utility.connection - INFO - Executing cmd: sudo sed -i '0,/.*:443 check$/s/.*:443 check$/        server 10.70.60.59 10.70.60.59:80 check\n&/' /etc/haproxy/haproxy.conf on 10.70.60.45
20:09:20 - MainThread - ocs_ci.utility.connection - INFO - Executing cmd: sudo sed -i '0,/.*:443 check$/s/.*:443 check$/        server 10.70.60.49 10.70.60.49:80 check\n&/' /etc/haproxy/haproxy.conf on 10.70.60.45
20:09:20 - MainThread - ocs_ci.utility.connection - INFO - Executing cmd: sudo sed -i '0,/.*:443 check$/s/.*:443 check$/        server 10.70.60.60 10.70.60.60:80 check\n&/' /etc/haproxy/haproxy.conf on 10.70.60.45
20:09:21 - MainThread - ocs_ci.utility.load_balancer - INFO - Restarting haproxy service on load balancer
20:09:21 - MainThread - ocs_ci.utility.connection - INFO - Executing cmd: sudo systemctl restart haproxy.service on 10.70.60.45
20:11:38 - MainThread - ocs_ci.utility.csr - INFO - Retrieving CSR data
20:11:38 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n default get csr  -n default -o yaml
20:11:39 - MainThread - ocs_ci.utility.csr - WARNING - Some nodes are not generated CSRs. Expected 10 but found 7 CSRs.retrying again
20:11:39 - MainThread - ocs_ci.utility.csr - INFO - Retrieving CSR data
20:11:39 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n default get csr  -n default -o yaml
20:11:41 - MainThread - ocs_ci.utility.utils - INFO - Going to sleep for 10 seconds before next iteration
20:11:51 - MainThread - ocs_ci.utility.csr - INFO - Retrieving CSR data


20:18:56 - MainThread - ocs_ci.utility.csr - INFO - Retrieving CSR data
20:18:56 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n default get csr  -n default -o yaml
20:18:57 - MainThread - ocs_ci.utility.csr - INFO - CSR generated for all 10 nodes
20:18:57 - MainThread - ocs_ci.utility.csr - INFO - Retrieving CSR data
20:18:57 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n default get csr  -n default -o yaml
20:18:58 - MainThread - ocs_ci.utility.csr - INFO - Pending CSRs: ['csr-fdsgk']
20:18:58 - MainThread - ocs_ci.utility.csr - INFO - Approving pending CSRs
20:18:58 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc adm certificate approve csr-fdsgk
20:18:59 - MainThread - ocs_ci.utility.csr - INFO - Checking for Pending CSRs
20:18:59 - MainThread - ocs_ci.utility.csr - INFO - Retrieving CSR data
20:18:59 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n default get csr  -n default -o yaml
20:19:01 - MainThread - ocs_ci.utility.csr - INFO - All CSRs approved
20:19:01 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc get Node  --selector=node-role.kubernetes.io/worker -o yaml
20:19:03 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc get Node  --selector=node-role.kubernetes.io/worker -o yaml
20:19:04 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc get Node  --selector=node-role.kubernetes.io/worker -o yaml
20:19:05 - MainThread - ocs_ci.ocs.node - INFO - Waiting for nodes ['compute-0', 'compute-1', 'compute-2', 'compute-3', 'compute-4', 'compute-5'] to reach status Ready
20:19:05 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc get node  -o yaml

20:20:30 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc get Node  -o yaml
20:20:32 - MainThread - ocs_ci.ocs.node - INFO - Node compute-4 reached status Ready
20:20:32 - MainThread - ocs_ci.ocs.node - INFO - The following nodes reached status Ready: ['compute-0', 'compute-1', 'compute-2', 'compute-3', 'compute-4', 'compute-5']
20:20:32 - MainThread - ocs_ci.ocs.node - INFO - New spun nodes: ['compute-4', 'compute-3', 'compute-5']
20:20:32 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc label node compute-4 cluster.ocs.openshift.io/openshift-storage=''
20:20:33 - MainThread - root - INFO - Successfully labeled compute-4 with OCS storage label
20:20:33 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc label node compute-3 cluster.ocs.openshift.io/openshift-storage=''
20:20:34 - MainThread - root - INFO - Successfully labeled compute-3 with OCS storage label
20:20:34 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc label node compute-5 cluster.ocs.openshift.io/openshift-storage=''
20:20:35 - MainThread - root - INFO - Successfully labeled compute-5 with OCS storage label
20:20:35 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc get Node  --selector=node-role.kubernetes.io/worker -o yaml
20:20:37 - MainThread - tests.manage.z_cluster.cluster_expansion.test_node_expansion - INFO - The worker nodes number after expansion 6


20:22:36 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage rsh rook-ceph-tools-67dfbfb8f5-mmqx8 ceph status --format json-pretty
20:22:40 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage rsh rook-ceph-tools-67dfbfb8f5-mmqx8 ceph health --format json-pretty
20:22:43 - MainThread - ocs_ci.ocs.cluster - INFO - {'checks': {}, 'status': 'HEALTH_OK'}
20:22:43 - MainThread - ocs_ci.ocs.cluster - INFO - [{'state_name': 'active+clean', 'count': 272}]
20:22:43 - MainThread - root - INFO - Re-balance is completed
PASSED                                                                                                                                                                [100%]
```

Nodes after running test case
```
(venv1) [vavuthu@localhost ocs-ci](add_node_vsphere)$ oc get nodes
NAME              STATUS   ROLES    AGE     VERSION
compute-0         Ready    worker   86m     v1.20.0+ba45583
compute-1         Ready    worker   86m     v1.20.0+ba45583
compute-2         Ready    worker   87m     v1.20.0+ba45583
compute-3         Ready    worker   6m40s   v1.20.0+ba45583
compute-4         Ready    worker   6m20s   v1.20.0+ba45583
compute-5         Ready    worker   6m41s   v1.20.0+ba45583
control-plane-0   Ready    master   109m    v1.20.0+ba45583
control-plane-1   Ready    master   110m    v1.20.0+ba45583
control-plane-2   Ready    master   109m    v1.20.0+ba45583
(venv1) [vavuthu@localhost ocs-ci](add_node_vsphere)$ 

```
Full logs: http://pastebin.test.redhat.com/938882

File: ocs_ci/ocs/platform_nodes.py [hunk 1]
```diff
@@ -1513,66 +1525,113 @@ def add_node(self):
...
+        logger.debug(f"Compute IP's after adding new nodes: {ips_after_adding_nodes}")
+        new_node_ips = list(set(ips_after_adding_nodes) - set(ips_before_adding_nodes))
+        logger.info(f"Newly added compute IP's: {new_node_ips}")
+
+        # inform load balancer regarding newly added nodes
+        lb = LoadBalancer()
+        lb.update_haproxy_with_nodes(new_node_ips)
+        lb.restart_haproxy()
+
+    def add_nodes_without_terraform(self):
```
Comment (vasukulkarni) [inline ocs_ci/ocs/platform_nodes.py:1581]: how about just add_nodes and update the docstring that it uses pyvmomi?

Comment (vavuthu) [reply ocs_ci/ocs/platform_nodes.py:1581]: i am using add_nodes in https://github.com/red-hat-storage/ocs-ci/pull/3825/files#diff-478e5b871f475f1e312384c0eeb216726f11da23e3d3ac4df80aa3ee5d92e441R1507

**Review by petr-balogh — APPROVED** [review]

**Review by sidhant-agrawal — APPROVED** [review]

File: ocs_ci/ocs/platform_nodes.py (hunk 1 above)
Comment (vavuthu) [reply ocs_ci/ocs/platform_nodes.py:1581]: @vasukulkarni let me know if you have any objections or improvements, I will take up in followup PR if needed

Comment (vasukulkarni) [reply ocs_ci/ocs/platform_nodes.py:1581]: @vavuthu sounds good, looks fine since its already used.

**Pull request merged**